File: py/hook/dump/Dump.py
import frida
import sys
import time

PACKAGE = "com.oakever.jigsawcard"
SAVE_FILE = "../metadata_dump.bin"

print("[*] connecting usb device...")

device = frida.get_usb_device(timeout=10)

front = device.get_frontmost_application()
print("[*] front app:", front)

pid = front.pid

print("[*] attaching pid:", pid)

session = device.attach(pid)

script_code = """
rpc.exports = {};

rpc.exports.dumpmeta = function () {

    var ranges = null;

    // Android frida runtime compatibility
    if (Process.enumerateRangesSync) {
        ranges = Process.enumerateRangesSync('r--');
    } else {
        ranges = Process.enumerateRanges('r--');
    }

    console.log("[JS] scan ranges:", ranges.length);

    for (var i = 0; i < ranges.length; i++) {

        var r = ranges[i];

        // metadata usually > 1MB
        if (r.size < 0x100000)
            continue;

        try {

            var data = Memory.readByteArray(r.base, r.size);
            if (!data) continue;

            var u8 = new Uint8Array(data);

            for (var j = 0; j < u8.length - 4; j++) {

                // global-metadata magic
             

                    var realBase = r.base.add(j);

                    console.log("[JS] FOUND metadata");
                    console.log("[JS] range base:", r.base);
                    console.log("[JS] offset:", j);
                    console.log("[JS] real base:", realBase);

                    var size = r.size - j;

                    return {
                        base: realBase.toString(),
                        size: size,
                        data: Memory.readByteArray(realBase, size)
                    };
                 
            }

        } catch (e) {
            // ignore unreadable range
        }
    }

    return null;
}
"""

print("[*] loading frida script...")

script = session.create_script(script_code)
script.load()

# important: wait unity load metadata
print("[*] waiting unity initialize...")
time.sleep(3)

api = script.exports_sync

print("[*] scanning memory...")

result = api.dumpmeta()

if result is None:
    print("[-] metadata not found")
    sys.exit(0)

print("[+] metadata FOUND")
print("[+] base :", result["base"])
print("[+] size :", result["size"])

data = bytes(result["data"])

with open(SAVE_FILE, "wb") as f:
    f.write(data)

print("[+] dump success:", SAVE_FILE)
print("[+] DONE")

time.sleep(2)

File: resources/jiacard/dump.js
function dumpIl2cpp() {

    var module = Process.findModuleByName("libil2cpp.so");
    if (!module) {
        console.log("[-] libil2cpp.so not found");
        return;
    }

    console.log("[+] Found libil2cpp.so");
    console.log("    Base:", module.base);
    console.log("    Size:", module.size);

    var ranges = module.enumerateRanges("r--");

    console.log("[+] Readable ranges:", ranges.length);

    var totalSize = 0;
    ranges.forEach(function (r) {
        totalSize += r.size;
    });

    var finalBuffer = new Uint8Array(totalSize);
    var offset = 0;

    ranges.forEach(function (range) {

        try {
            var bytes = range.base.readByteArray(range.size);
            var arr = new Uint8Array(bytes);

            finalBuffer.set(arr, offset);
            offset += arr.length;

        } catch (e) {
            console.log("[-] Skip unreadable range:", range.base);
        }

    });

    console.log("[+] Sending to PC...");
    send("done", finalBuffer.buffer);
    console.log("[✓] Dump sent.");
}

setImmediate(function () {

    var timer = setInterval(function () {

        var m = Process.findModuleByName("libil2cpp.so");
        if (m) {
            clearInterval(timer);
            dumpIl2cpp();
        }

    }, 500);
});
